alc-verb/main.c

```diff
@@ -0,0 +1,244 @@
+{
+	int c = 0;
+	
+	for (; tbl->str; tbl++)
+	{
+		unsigned long len = strlen(tbl->str) + 2;
+		
+		if (!one_per_line && c + len >= 80)
+		{
+			fprintf(stderr, "\n");
+			c = 0;
+		}
+		
+		if (one_per_line)
+			fprintf(stderr, "  %s\n", tbl->str);
+		else if (!c)
+			fprintf(stderr, "  %s", tbl->str);
+		else
+			fprintf(stderr, ", %s", tbl->str);
+		
+		c += 2 + len;
+	}
+	
+	if (!one_per_line)
+		fprintf(stderr, "\n");
+}
+
+/* look up a value from the given string table */
+static int lookup_str(struct strtbl *tbl, const char *str)
+{
+	struct strtbl *p, *found;
+	unsigned long len = strlen(str);
+	
+	found = NULL;
+	
+	for (p = tbl; p->str; p++)
+	{
+		if (!strncmp(str, p->str, len))
+		{
+			if (found)
+			{
+				fprintf(stderr, "No unique key '%s'\n", str);
+				return -1;
+			}
+			
+			found = p;
+		}
+	}
+	
+	if (!found)
+	{
+		fprintf(stderr, "No key matching with '%s'\n", str);
+		return -1;
+	}
+	
+	return found->val;
+}
+
+/* convert a string to upper letters */
+static void strtoupper(char *str)
+{
+	for (; *str; str++)
+		*str = toupper(*str);
+}
+
+static void usage(void)
+{
+	fprintf(stderr, "alc-verb for AppleALC (based on alsa-tools hda-verb)\n");
+	fprintf(stderr, "usage: alc-verb [option] nid verb param\n");
+	fprintf(stderr, "   -l      List known verbs and parameters\n");
+	fprintf(stderr, "   -L      List known verbs and parameters (one per line)\n");
+}
+
+static void list_verbs(int one_per_line)
+{
+	fprintf(stderr, "known verbs:\n");
+	list_keys(hda_verbs, one_per_line);
+	fprintf(stderr, "known parameters:\n");
+	list_keys(hda_params, one_per_line);
+}
+
+int main(int argc, char **argv)
+{
+	long nid, verb, params;
+	int c;
+	char **p;
+	bool quiet = false;
+	
+	while ((c = getopt(argc, argv, "qlL")) >= 0)
+	{
+		switch (c)
+		{
+			case 'l':
+				list_verbs(0);
+				return 0;
+			case 'L':
+				list_verbs(1);
+				return 0;
+			case 'q':
+				quiet = true;
+				break;
+			default:
+				usage();
+				return 1;
+		}
+	}
+	
+	if (argc - optind < 3)
+	{
+		usage();
+		return 1;
+	}
+	
+	p = argv + optind;
+	nid = strtol(*p, NULL, 0);
+	if (nid < 0 || nid > 0xff) {
+		fprintf(stderr, "invalid nid %s\n", *p);
+		return 1;
+	}
+	
+	p++;
+	if (!isdigit(**p))
+	{
+		strtoupper(*p);
+		verb = lookup_str(hda_verbs, *p);
+		
+		if (verb < 0)
+			return 1;
+	}
+	else
+	{
+		verb = strtol(*p, NULL, 0);
+		
+		if (verb < 0 || verb > 0xfff)
+		{
+			fprintf(stderr, "invalid verb %s\n", *p);
+			return 1;
+		}
+	}
+	
+	p++;
+	if (!isdigit(**p))
+	{
+		strtoupper(*p);
+		params = lookup_str(hda_params, *p);
+		if (params < 0)
+			return 1;
+	}
+	else
+	{
+		params = strtol(*p, NULL, 0);
+		
+		if (params < 0 || params > 0xffff)
+		{
+			fprintf(stderr, "invalid param %s\n", *p);
+			return 1;
+		}
+	}
+
+	if (!quiet)
+		printf("nid = 0x%lx, verb = 0x%lx, param = 0x%lx\n", nid, verb, params);
+	
+	// Execute command
+	uint32_t result = execute_command(nid, verb, params);
+
+	// Print result
+	printf("0x%08x\n", result);
```

Comment:
Better cast that to `(unsigned)`. On some architectures `uint32_t` is not `unsigned int` but rather `unsigned long`.

Reply:
Didn't know that. Resolved.